### PR TITLE
Arc-Length Surface Loss Reweighting: fix non-uniform mesh density bias

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The surface MAE loss weights every surface node equally, but our mesh has non-uniform node density — nodes cluster at the leading edge (LE) and trailing edge (TE), while the mid-chord region has coarser spacing. This means LE/TE nodes dominate the gradient signal even if the model already predicts them well. Mid-chord errors (where tandem wake interaction manifests) are under-penalized.

The fix: reweight each surface node's loss contribution by its arc-length element — the chord-normalized half-distance to its two nearest surface neighbors. This makes the training loss topology-consistent with the evaluation metric (which evaluates uniformly over all nodes), and concentrates gradient signal on the physically important mid-chord and suction-peak regions where OOD tandem errors are largest.

**Why now:** The DCT frequency-weighted loss (PR #2184, -0.3% p_tan, merged) works in the frequency domain to up-weight high-frequency features. Arc-length reweighting works in the spatial domain on the mesh quadrature correction. The two are orthogonal and complementary — both can be active simultaneously.

**Key insight:** A node at x/c=0.02 with 5 neighbors contributes ~5x more to the current loss than a node at x/c=0.5 with 1 neighbor. But p_tan (our evaluation metric) counts each node equally. Correcting this eliminates a systematic mismatch between training signal and evaluation.

### Step 1: Compute arc-length weights (once, at dataset init or start of training)

In `train.py` or a preprocessing function, compute per-surface-node arc-length weights:

```python
def compute_arclength_weights(surface_coords):
    from scipy.spatial import cKDTree
    import numpy as np
    
    tree = cKDTree(surface_coords)
    dists, _ = tree.query(surface_coords, k=3)  # k=3: self + 2 neighbors
    dists = dists[:, 1:]  # exclude self: shape (N_surf, 2)
    ds = dists.mean(axis=1)  # arc-length element per node
    w = ds / (ds.mean() + 1e-8)  # normalize so mean = 1.0
    w = np.clip(w, 0.1, 10.0)   # clamp degenerate nodes
    return torch.tensor(w, dtype=torch.float32)
```

**Cache these weights** — compute once per sample during dataset loading, not per forward pass. Surface nodes are identified by boundary IDs 5, 6, 7.

### Step 2: Apply weights in the surface loss computation

Find the surface MAE computation in `train.py` and apply per-node weights:

```python
# Before (equal weighting):
surf_loss = (pred[surf_mask] - target[surf_mask]).abs().mean()

# After (arc-length weighted):
surf_w = arclength_weights[surf_mask]        # (N_surf,) weights
surf_loss = ((pred[surf_mask] - target[surf_mask]).abs() * surf_w.unsqueeze(-1)).mean()
```

Only apply to surface nodes. Do NOT apply to volume nodes or the full-domain loss.

### Step 3: Add flag and logging

Add `--arclength_surface_loss` boolean flag. Log the weight statistics to W&B for the first batch:
`wandb.log({"arclength_w_mean": surf_w.mean(), "arclength_w_max": surf_w.max(), "arclength_w_min": surf_w.min()})` — confirms weights are reasonable (should be ~1.0 mean, 3-8x ratio between TE-dense and mid-chord nodes).

### Step 4: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent fern --seed 42 \
  --wandb_name "fern/arclength-surf-s42" \
  --wandb_group "fern/arclength-surface-loss" \
  --arclength_surface_loss \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5

# Seed 73: same flags, --seed 73 --wandb_name "fern/arclength-surf-s73"
```

Note: `--dct_freq_loss` is intentionally included — run alongside arc-length reweighting since they are orthogonal corrections (frequency vs spatial domain).

## Baseline

Current best (PR #2184, DCT Frequency-Weighted Loss, 2-seed avg):

| Metric | Baseline | Target |
|--------|---------|--------|
| p_in | 13.205 | < 13.21 |
| p_oodc | 7.816 | < 7.82 |
| **p_tan** | **28.502** | **< 28.50** |
| p_re | 6.453 | < 6.45 |

W&B baseline runs: `6yfv5lio` (seed 42, p_tan=28.432), `etepxvjc` (seed 73, p_tan=28.572)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Arc-Length Surface Loss Reweighting

### Setup
- **Branch:** `fern/arclength-surface-loss-reweight`
- **Flag:** `--arclength_surface_loss` (+ all baseline flags + DCT freq loss)
- **Epochs:** 149 (180-min timeout)
- **W&B runs (offline):** `cor600he` (s42), `5pfgo0bx` (s73)

### Results — Surface Pressure MAE (lower is better)

| Split | Baseline s42 | Baseline s73 | Baseline avg | Arclength s42 | Arclength s73 | Arclength avg | Δ |
|-------|-------------|-------------|-------------|---------------|---------------|---------------|---|
| p_in | 13.205 | — | 13.205 | 14.08 | 14.46 | 14.27 | +8.1% ❌ |
| p_tan | 28.502 | — | 28.502 | 28.3 | 29.6 | 28.95 | +1.6% ❌ |
| p_oodc | 7.816 | — | 7.816 | 8.1 | 7.8 | 7.95 | +1.7% ❌ |
| p_re | 6.453 | — | 6.453 | 6.6 | 6.4 | 6.50 | +0.7% ❌ |

### Arc-Length Weight Statistics (logged at step 0)
- Mean: 1.0 (normalized)
- Min: 0.094 (dense LE/TE nodes downweighted ~10x)
- Max: 10.35 (sparse mid-chord nodes upweighted ~10x)

### Analysis

**Arc-length surface loss reweighting did NOT beat baseline.** All splits are slightly worse.

**Why it likely didn't help:**

1. **Existing hard-node mining already addresses the density bias.** The hard-node mining mechanism (activated at epoch 30) upweights high-error nodes, which naturally concentrates on LE/TE regions where pressure gradients are steepest. Arc-length reweighting actively *downweights* these dense LE/TE nodes, working against the hard-node mining signal.

2. **LE/TE density is a feature, not a bug.** The mesh is deliberately refined at LE/TE because that's where the flow physics are most complex (stagnation points, separation, suction peaks). Downweighting these critical regions during training reduces the model's ability to learn the hardest parts of the flow.

3. **Cross-foil contamination in tandem cases.** The KNN-based arc-length computation uses Euclidean nearest neighbors across all surface nodes. For tandem configurations, a TE node on foil 1 could pick up an LE node on foil 2 as a "neighbor," producing incorrect weights. This affects the tandem transfer split specifically.

4. **Wide clamp range [0.1, 10.0] may be too aggressive.** The 10x upweighting of sparse mid-chord nodes significantly distorts the loss landscape. A tighter range like [0.2, 5.0] might produce more modest effects.

### Suggested Follow-ups

- **Not recommended to pursue further** — the theoretical motivation is undermined by the existing hard-node mining mechanism. The mesh density at LE/TE is intentional and beneficial for training.
- If revisited, could try: (1) tighter clamp [0.2, 3.0], (2) per-foil arc-length computation using boundary IDs to avoid cross-foil contamination, (3) applying only to volume nodes (not surface).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2210 because ADVISOR: Closing — arc-length surface loss reweighting conflicts structurally with hard-node mining, which is already a core part of our training pipeline. The two mechanisms have opposing intent: arc-length upweights mid-chord (where mesh is sparse) while hard-node mining upweights LE/TE (where errors are highest). They cancel each other out, and the results confirm this clearly.

Results vs new baseline (PR #2213, p_in=11.979, p_oodc=7.643, p_tan=28.341, p_re=6.300):
- p_in: 14.27 (+19.1% regression)
- p_oodc: 7.95 (+4.1%)
- p_tan: 28.95 (+2.1%)
- p_re: 6.50 (+3.2%)

All four metrics regress significantly. The arc-length weights (min=0.094 at LE/TE, max=10.35 at mid-chord) effectively suppress gradient signal from the most information-rich regions for pressure prediction.

Your root-cause analysis is correct. The hypothesis was physically motivated but not compatible with the existing loss structure without disabling hard-node mining first — which itself improves performance and is not worth removing. Good work diagnosing this cleanly..
